src/adapter/src/coord/read_policy.rs

```diff
@@ -252,27 +324,75 @@ impl<S: Append + 'static> crate::coord::Coordinator<S> {
         self.read_capability.remove(id).is_some()
     }
 
-    /// Acquire read holds on the indicated collections at the indicated time.
+    fn initialize_read_holds(
+        &mut self,
+        time: mz_repr::Timestamp,
+        id_bundle: CollectionIdBundle,
+    ) -> ReadHolds<mz_repr::Timestamp> {
+        let mut read_holds = ReadHolds::new();
+
+        for id in id_bundle.storage_ids.iter() {
+            let collection = self.controller.storage.collection(*id).unwrap();
+            let read_frontier = collection.read_capabilities.frontier();
+            let time = if read_frontier.less_equal(&time) {
+                time
+            } else {
+                read_frontier
+                    .to_owned()
+                    .first()
+                    .owned()
+                    .unwrap_or(mz_repr::Timestamp::MAX)
```

Comment:
This doesn't seem right, is there a better way of doing this? This is one of the points where we no longer work with Antichains and force it into a `u64`.

Reply:
Is there a reason to not use `ReadHolds<Antichain<mz_repr::Timestamp>>`? MAX might be wrong here because `Antichain([MAX]) < Antichain([])`, so we're incorrectly saying some timestamp exists when we know it doesn't.

Why wasn't this a problem before? Because we were never comparing `collection.read_frontier()` to a ReadHold?

Reply:
I'll play around with this change it sounds like a better idea. We'd only ever compare the read hold via something like `collection.read_frontier().less_equal(timestamp_of_type_mz_repr::Timestamp)`, which is why we never ran into a problem.

Reply:
Ok, I just pushed a commit that changes the `ReadHolds` implementation to use `Antichains`. I think it actually nicely matches the implementation of `ReadCapability` and `ReadPolicy`.

Reply:
> I'm not sure if using an antichain is going to work here or not. Conceptually, if the collection frontier is empty, it would mean there's no possible Timestamp at which we can hold a read. Can/should we return an error or do nothing in that case? Just trying to think about this as a concept first.

@frankmcsherry do you have any thoughts on this? We talked about this briefly last week.

If the collection frontier is empty, then we can acquire a read hold on the empty frontier. The read hold timestamp eventually gets converted into a `ReadPolicy::ValidFrom` type, which is implemented using an `Antichain` (https://github.com/MaterializeInc/materialize/blob/84bea37fc8575c1753a5a2a85fb853c05e8eb3ef/src/storage/src/controller.rs#L356), so the conversion ends up being pretty easy.

However, the read hold would be pretty useless, but at least it provides some bookkeeping for that object. Also, I think if an object has been compacted to the empty frontier it probably shouldn't be the read hold API's job to deal with it.

Reply:
I spoke with @frankmcsherry  in person about this, it's fine for book keeping purpose to include empty frontiers in the read holds as long as storage and compute don't freak out when we try request the read holds. So I took a look at those functions, it seems like if we pass an empty frontier, then it will be a no-op.
https://github.com/MaterializeInc/materialize/blob/15272a8c6fb9db52711c969359a598183fcc091a/src/compute-client/src/controller/instance.rs#L607-L648
https://github.com/MaterializeInc/materialize/blob/15272a8c6fb9db52711c969359a598183fcc091a/src/storage/src/controller.rs#L1217-L1239